In [ ]:
!pip install -U transformers

## Local Inference on GPU
Model page: https://huggingface.co/openai/whisper-large-v3

⚠️ If the generated code snippets do not work, please open an issue on either the [model repo](https://huggingface.co/openai/whisper-large-v3)
			and/or on [huggingface.js](https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/model-libraries-snippets.ts) 🙏

In [ ]:
# Use a pipeline as a high-level helper
from transformers import pipeline

pipe = pipeline("automatic-speech-recognition", model="openai/whisper-large-v3")

config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/3.09G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/1259 [00:00<?, ?it/s]

generation_config.json: 0.00B [00:00, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

normalizer.json: 0.00B [00:00, ?B/s]

added_tokens.json: 0.00B [00:00, ?B/s]

special_tokens_map.json: 0.00B [00:00, ?B/s]

preprocessor_config.json:   0%|          | 0.00/340 [00:00<?, ?B/s]

In [ ]:
# Load model directly
from transformers import AutoProcessor, AutoModelForSpeechSeq2Seq

processor = AutoProcessor.from_pretrained("openai/whisper-large-v3")
model = AutoModelForSpeechSeq2Seq.from_pretrained("openai/whisper-large-v3")

Loading weights:   0%|          | 0/1259 [00:00<?, ?it/s]

## Remote Inference via Inference Providers
Ensure you have a valid **HF_TOKEN** set in your environment. You can get your token from [your settings page](https://huggingface.co/settings/tokens). Note: running this may incur charges above the free tier.
The following Python example shows how to run the model remotely on HF Inference Providers, automatically selecting an available inference provider for you.
For more information on how to use the Inference Providers, please refer to our [documentation and guides](https://huggingface.co/docs/inference-providers/en/index).

In [ ]:
import os
os.environ['HF_TOKEN'] = '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'

In [ ]:
import os
from huggingface_hub import InferenceClient

client = InferenceClient(
    provider="auto",
    api_key=os.environ["HF_TOKEN"],
)

output = client.automatic_speech_recognition("/content/sample_audio.wav", model="openai/whisper-large-v3")

HfHubHTTPError: Client error '402 Payment Required' for url 'https://router.huggingface.co/fal-ai/fal-ai/whisper' (Request ID: Root=1-69c94387-385e8e2d7ee79d4332d5f0c1;c25fe5c2-80e3-477b-b488-e25fd0f808e0)
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/402

Pre-paid credits are required to use provider fal-ai. Add pre-paid credits to your account to start using fal-ai.

### Testing Local Inference with Pipeline
Let's use the `pipe` object created earlier to transcribe the sample audio file.

In [ ]:
output_pipeline = pipe('/content/sample_audio.wav')
print(output_pipeline)

### Testing Local Inference with Direct Model Load
Now, let's use the `processor` and `model` objects for transcription. This requires a few more steps:
1.  Load the audio file.
2.  Process the audio using the `processor`.
3.  Generate predictions using the `model`.
4.  Decode the predictions back to text using the `processor`.

In [ ]:
import torch
import librosa

# Check for GPU and set device
device = "cuda" if torch.cuda.is_available() else "cpu"

# Move model to device and cast to float16
model.to(device).half() # .half() converts the model's parameters to float16

# Load the audio file
audio_input, sample_rate = librosa.load('/content/sample_audio.wav', sr=16000)

# Process the audio
input_features = processor(audio_input, sampling_rate=sample_rate, return_tensors="pt").input_features

# Move input features to device and cast to float16
input_features = input_features.to(device).half()

# Generate predictions
predicted_ids = model.generate(input_features)

# Decode predictions to text
transcription = processor.batch_decode(predicted_ids, skip_special_tokens=True)[0]
print(f"Transcription (direct load): {transcription}")

The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
A custom logits processor of type <class 'transformers.generation.logits_process.SuppressTokensLogitsProcessor'> has been passed to `.generate()`, but it was also created in `.generate()`, given its parameterization. The custom <class 'transformers.generation.logits_process.SuppressTokensLogitsProcessor'> will take precedence. Please check the docstring of <class 'transformers.generation.logits_process.SuppressTokensLogitsProcessor'> to see related `.generate()` flags.
A custom logits processor of type <class 'transformers.generation.logits_process.SuppressTokensAtBeginLogitsProcessor'> has been passed to `.generate()`, but it was also created in `.generate()`, given its parameterization. The custom <class 'transformers.generation.logits_process.SuppressTokensA

Transcription (direct load):  तो बैंक में जाके पैसे लगाते थे और तब वह पैसा वहाँ से भेजता था फिर यहाँ से लोग जाते थे पैसे बैंक से निकाल कर ले आते थे अब है अब तुरंत हाथ वहाथ मिल जाता है पैसा और किसी से कोई समान खरीदे दोकान पर गए तो पैसे से मोबाइल से जो है पैसे दोकानदार को दे देते थे और समान हम लोगों को मिल जाता है


### Remote vs. Local Inference

**Remote Inference via Inference Providers (e.g., Hugging Face Inference Providers):**

*   As demonstrated in the 'Remote Inference via Inference Providers' section, using services like 'fal-ai' (automatically selected by `provider="auto"` in `InferenceClient`) requires **pre-paid credits** on your Hugging Face account. Attempting to use these services without sufficient credits will result in a `402 Payment Required` error. Ensure you have a valid `HF_TOKEN` set in your environment and sufficient credits if you wish to use these providers.

**Local Inference (using `transformers` library on Colab):**

*   The model can be run locally on your Colab instance (which often utilizes a GPU provided by Colab for free tier, or paid tier for more powerful GPUs). This method does **not** incur additional charges related to inference providers.
*   Two methods for local inference are showcased:
    1.  **Using a high-level pipeline:** As shown in cell `jxY_cNURK1DK` (initial setup) and cell `519b074d` (execution example). This is the simplest way to get started.
    2.  **Loading the model directly:** As shown in cell `-nlwy0IZK1DK` (initial setup) and cell `35c5a39f` (execution example). This method offers more control over the pre-processing and model execution steps. When using a GPU, it's often beneficial to cast the model and input features to `float16` for performance, as demonstrated in the corrected cell `35c5a39f`.